### DISCUSSION TITLE
Create or update in attributes

### DISCUSSION THREAD

**Discussion by vincent-lg** [evennia/evennia#2314] [Evennia Engine Development]
Hi,

On the Evennia chat I often find this question: when is `at_object_creation` called and why my attributes aren't present?  So I was thinking.  Attributes are great, they are flexible, but the fact they can be there (and can be absent) means you need to implement more logic, and there's no easy way to update an object to reset "default" values to attributes that do not exist.

So I suggest something on two levels.

First a method which, contrary to `at_object_creation`, is called every time the object is first loaded.  `__init__` could do the trick, but I don't think it's a good idea to override it.  And there's a `at_init` isn't there?  If so, nothing more to add for this step.

The second step would be in the `AttributeHandle`: add a method/property that behaves like a context manager.  I'll give you an example of usage:

```
    def at_init(self):
        with self.db.if_necessary as update:
            update.attr1 = 25
            update.attr2 = 35
```

In this context, `attr1` and `attr2` will only be set if they don't exist in the database.  It would make updating much easier while still playing with attributes.  Of course, the context manager name (and the way updates are handled) could be different, that's just a suggestion.

To be debated!

**Comment by Griatch** [comment]
Hm, not sure I follow the advantage of the context manager here - in your example it's doing the same work as 

```
def at_init(self):
    if not (self.db.attr1 and self.db.attr2):
        self.db.attr1 = 25
        self.db.attr2 = 35
```

In fact, the manager would have to do exactly this check as well, but hidden inside the manager. Isn't this simple `if` statement a lot more explicit and readable than hiding this logic in a manager you'll only use once?

**Reply by vincent-lg** [reply]
Maybe.  However, doing something like `if not self.db.attr1 or not self.db.attr2` might lead to errors.  In this context, what if `attr2` has been defined but `attr1` hasn't?  Both will go back to default values and `attr2` will be replaced.  The same applies if you try to change the condition, though in reverse.  Of course with more attributes the condition becomes even more complicated.  So I don't know: easier to read?  Perhaps.  But easier to make mistakes?  Probably so.  The full syntax (`with ... as ...`) might not be necessary here, if that's too confusing:

```
    def at_init(self):
        with self.db.if_necessary:
            self.db.attr1 = ...
```

This might be easier to read, I don't know.

**Reply by Griatch** [reply]
Hm, I guess your second example would not work, because that's literally just an if statement hidden away. Your first version uses the handler to check (per-attribute, I suppose), which should be set.

It may be easier to visualize with a different name:

```
with update_if_none as update:
    update.foo = 1
    update.bar = 1
```

That said, I don't think this makes things easier than the longer but a lot more explicit
```
if not self.db.foo:
    self.db.foo = 1
if not self.db.bar:
    self.db.bar = 1
```

**Reply by vincent-lg** [reply]
It's true that one would need to do that in a typeclass, like for rooms or exits, but I guess that's the case when using attributes anyway.

The idea is, if it's inside the `if_necessary` context manager, the attribute is only set if it doesn't exist.

So if you do something like:

```python
self.db.attr1 = 45
self.db.attr2 = 38

with self.db.if_necessary as default: # Let's call it default for the example
    default.attr1 = "something else" # won't be called, attr1 exists.
    default.attr2 = -12 # won't be called, attr2 exists.
    default.attr3 = "default" # attr3 is created, and set to this value.
```

In short, inside a `if_necessary` context manager, when writing an attribute, the `AttributeHandler` will check if the attribute exists already, and if it does, do nothing.  If it doesn't, write it with the specified value.

So this might have been called default attribute value, might have been clearer.  That was the idea.  I'm not sure how useful it is, but if you have a tight connection between typeclasses and the attributes you use (something quite common), it might become handy.

**Comment by davewiththenicehat** [comment]
I may be misunderstanding what you are doing, however:

You can override the handlers `__getattribute()__` and `__setattr()__`.
It requires a good grasp of object orientation. It requires a very good knowledge of descriptors. Especially `__getattribute()__`, I recommend not overriding `__getattr()__`. You will want to use `object.__getattribute()__` within your overridden `__getattribute()__`, because super() interaction here is real hard to follow.
If you want default values you do have to provide them. I use a dictionary for managed attributes that is checked again in `__getattribute()__`. The dictionary is short and specific to that object.
You can get a list of existing attributes from `self.__dict()__`.

Does AttributeHandler return False of None when an attribute does not exist? I ask because those if statements may need to be try blocks.

Hope that helps.

**Comment by strikaco** [comment]
> The second step would be in the AttributeHandle: add a method/property that behaves like a context manager. I'll give you an example of usage:

This is an odd use of a context manager, and doesn't make clear what the "context" actually _is_.

Just refactor the AttributeHandler and implement the interface of Django's cache. The `add()` method would--by default--set the attribute if it doesn't exist and leave it untouched otherwise. A new method (`set()`) would overwrite the current value.

(Caching was removed from this implementation; was to be re-added via mixin)

```
from abc import ABC, abstractmethod
from evennia.utils.dbserialize import to_pickle, from_pickle
from evennia.utils.utils import to_str

DEFAULT_TIMEOUT = 60 * 60 * 24

class BaseHandler(ABC):
    _missing_key = object()
    _version = 0

    def __init__(self, obj, *args, **kwargs):
        self.obj = obj
        self._objid = obj.id
        self._model = to_str(obj.__dbclass__.__name__.lower())
        
        super().__init__()

    def make_key(self, key, *args, version=None, **kwargs):
        """
        Construct the key used by all other methods. By default, use the
        key_func to generate a key (which, by default, prepends the
        `key_prefix' and 'version'). A different key function can be provided
        at the time of cache construction; alternatively, you can subclass the
        cache backend to provide custom key making behavior.
        """
        if version is None:
            version = self._version

        return (key, *args, version)
        
    @classmethod
    def pickle(cls, value, *args, **kwargs):
        return to_pickle(value)
        
    @classmethod
    def unpickle(cls, value, *args, **kwargs):
        return from_pickle(value)

    @abstractmethod
    def add(self, key, value, *args, **kwargs):
        """
        Set a value in the cache if the key does not already exist. If
        timeout is given, use that timeout for the key; otherwise use the
        default cache timeout.
        Return True if the value was stored, False otherwise.
        """
        raise NotImplementedError('subclasses of BaseCache must provide an add() method')

    @abstractmethod
    def get(self, key, *args, **kwargs):
        """
        Fetch a given key from the cache. If the key does not exist, return
        default, which itself defaults to None.
        """
        raise NotImplementedError('subclasses of BaseCache must provide a get() method')

    @abstractmethod
    def set(self, key, value, *args, **kwargs):
        """
        Set a value in the cache. If timeout is given, use that timeout for the
        key; otherwise use the default cache timeout.
        """
        raise NotImplementedError('subclasses of BaseCache must provide a set() method')

    @abstractmethod
    def touch(self, key, ttl, *args, **kwargs):
        """
        Update the key's expiry time using timeout. Return True if successful
        or False if the key does not exist.
        """
        raise NotImplementedError('subclasses of BaseCache must provide a touch() method')
        
    @abstractmethod
    def clear(self):
        """Remove *all* values from the cache at once."""
        raise NotImplementedError('subclasses of BaseCache must provide a clear() method')

    @abstractmethod
    def delete(self, key, *args, **kwargs):
        """
        Delete a key from the cache and return whether it succeeded, failing
        silently.
        """
        raise NotImplementedError('subclasses of BaseCache must provide a delete() method')
        
    @abstractmethod
    def all(self, *args, **kwargs):
        raise NotImplementedError('subclasses of BaseCache must provide an all() method')

    def get_many(self, keys, *args, version=None, **kwargs):
        """
        Fetch a bunch of keys from the cache. For certain backends (memcached,
        pgsql) this can be *much* faster when fetching multiple values.
        Return a dict mapping each key in keys to its value. If the given
        key is missing, it will be missing from the response dict.
        """
        d = {}
        for k in keys:
            val = self.get(k, self._missing_key, version=version)
            if val is not self._missing_key:
                d[k] = val
        return d

    def get_or_set(self, key, default, *args, timeout=DEFAULT_TIMEOUT, version=None, **kwargs):
        """
        Fetch a given key from the cache. If the key does not exist,
        add the key and set it to the default value. The default value can
        also be any callable. If timeout is given, use that timeout for the
        key; otherwise use the default cache timeout.
        Return the value of the key stored or retrieved.
        """
        val = self.get(key, self._missing_key, version=version)
        if val is self._missing_key:
            if callable(default):
                default = default()
            self.add(key, default, timeout=timeout, version=version)
            # Fetch the value again to avoid a race condition if another caller
            # added a value between the first get() and the add() above.
            return self.get(key, default, version=version)
        return val

    def has_key(self, key, *args, version=None, **kwargs):
        """
        Return True if the key is in the cache and has not expired.
        """
        return self.get(key, self._missing_key, version=version) is not self._missing_key

    def incr(self, key, delta=1, *args, version=None, **kwargs):
        """
        Add delta to value in the cache. If the key does not exist, raise a
        ValueError exception.
        """
        value = self.get(key, self._missing_key, version=version)
        if value is self._missing_key:
            raise ValueError("Key '%s' not found" % key)
        new_value = (value or 0) + delta
        self.set(key, new_value, version=version)
        return new_value

    def decr(self, key, delta=1, *args, version=None, **kwargs):
        """
        Subtract delta from value in the cache. If the key does not exist, raise
        a ValueError exception.
        """
        return self.incr(key, -delta, version=version)

    def __contains__(self, key):
        """
        Return True if the key is in the cache and has not expired.
        """
        # This is a separate method, rather than just a copy of has_key(),
        # so that it always has the same functionality as has_key(), even
        # if a subclass overrides it.
        return self.has_key(key)

    def set_many(self, data, *args, timeout=DEFAULT_TIMEOUT, version=None, **kwargs):
        """
        Set a bunch of values in the cache at once from a dict of key/value
        pairs.  For certain backends (memcached), this is much more efficient
        than calling set() multiple times.
        If timeout is given, use that timeout for the key; otherwise use the
        default cache timeout.
        On backends that support it, return a list of keys that failed
        insertion, or an empty list if all keys were inserted successfully.
        """
        for key, value in data.items():
            self.set(key, value, timeout=timeout, version=version)
        return []

    def delete_many(self, keys, *args, version=None, **kwargs):
        """
        Delete a bunch of values in the cache at once. For certain backends
        (memcached), this is much more efficient than calling delete() multiple
        times.
        """
        for key in keys:
            self.delete(key, *args, version=version, **kwargs)

from evennia.typeclasses.attributes import Attribute

class AttributeHandler(BaseHandler):

    _m2m_fieldname = "db_attributes"
    _attrtype = None
    _version = 1

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        query = {
            "%s__id" % self._model: self._objid,
            "attribute__db_model__iexact": self._model,
            "attribute__db_attrtype": self._attrtype,
        }
        self.qs = getattr(self.obj, self._m2m_fieldname).through.objects.filter(**query).order_by('id')
        # q.attribute for q in self.qs.all()

    def all(self, *args, **kwargs):
        return tuple(x.attribute for x in self.qs.all())

    def get(self, key, *args, category=None, return_obj=False, return_iter=False, **kwargs):
        """
        Fetch a given key from the cache. If the key does not exist, return
        default, which itself defaults to None.
        """
        query = {
            "attribute__db_key__iexact": key.lower(),
            "attribute__db_category__iexact": category.lower() if category else None,
        }
        attrs = (x.attribute for x in self.qs.filter(**query))
        if return_obj:
            attrs = attrs
        else: attrs = (x.value or x.db_strvalue for x in attrs)
        
        attrs = tuple(attrs)
        if return_iter: return attrs
        if attrs: return attrs[0]
        return None

    def add(self, key, value, *args, **kwargs):
        """
        Set a value in the cache if the key does not already exist. If
        timeout is given, use that timeout for the key; otherwise use the
        default cache timeout.
        Return True if the value was stored, False otherwise.
        """
        if self.get(key, *args, **kwargs): return False

        self.set(key, value, *args, **kwargs)
        return True

    def set(self, key, value, *args, category=None, strattr=False, **kwargs):
        """
        Set a value in the cache. If timeout is given, use that timeout for the
        key; otherwise use the default cache timeout.
        """
        # Try updating existing
        existing = self.get(key, *args, **kwargs)
        if existing:
            attr = existing[0]
            if strattr:
                # store as a simple string (will not notify OOB handlers)
                attr.db_strvalue = value
                attr.save(update_fields=["db_strvalue"])
            else:
                # store normally (this will also notify OOB handlers)
                attr.value = value
        else:
            kwargs = {
                "db_key": key,
                "db_category": category,
                "db_model": self._model,
                "db_attrtype": self._attrtype,
                "db_value": None if strattr else self.pickle(value),
                "db_strvalue": value if strattr else None,
            }
            attr = Attribute(**kwargs)
            attr.save()
            getattr(self.obj, self._m2m_fieldname).add(attr)

    def delete(self, key, *args, **kwargs):
        """
        Delete a key from the cache and return whether it succeeded, failing
        silently.
        """
        attrs = self.get(key, *args, **kwargs)
        if attrs:
            return bool([x.delete() for x in attrs])
        return False

    def clear(self, *args, **kwargs):
        for attr in self.all():
            attr.delete()

    def touch(self, *args, **kwargs):
        pass


from evennia.utils.test_resources import EvenniaTest

class AttributeHandlerTest(EvenniaTest):

    def test_stuff(self):

        hnd = AttributeHandler(self.char1)
        self.assertTrue(hnd.add('bozo', 'clown'))
        self.assertFalse(hnd.add('bozo', 'office-guy'))
        
        self.assertEqual(hnd.get('bozo'), 'clown')
        
        self.assertEqual(hnd.incr('dubs', 1), 1)
`
